### PR TITLE
Enable basket export for kek

### PR DESCRIPTION
**Pull request by domi4484**
Based on #182

### CODE REVIEW

**Comment by ponceta** [comment]
https://github.com/QGEP/qgepqwat2ili/pull/181 will be rebased after this.

**Comment by sjib** [comment]
```
======================================================================
ERROR: test_case_b_export_complete_qgep_to_xtf (qgepqwat2ili.tests.test_qgep.TestQGEPUseCases.test_case_b_export_complete_qgep_to_xtf)
# B. export the whole QGEP model to INTERLIS
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1256, in _execute_context
    self.dialect.do_executemany(
  File "/opt/hostedtoolcache/Python/3.12.7/x64/lib/python3.12/site-packages/sqlalchemy/dialects/postgresql/psycopg2.py", line 912, in do_executemany
    cursor.executemany(statement, parameters)
psycopg2.errors.NotNullViolation: null value in column "t_basket" violates not-null constraint
DETAIL:  Failing row contains (0, null, organisation, ch13p7mzOG000002).
```

@domi4484 Can you fix this also?

**Comment by sjib** [comment]
@domi4484 It looks like organisation also needs a t_basket column. Could you finish this asap - there are other developments waiting for this to be finished.

**Comment by domi4484** [comment]
@sjib the basket column get created. But for some reason sometimes the value does not get populated.

File: qgepqwat2ili/utils/qgep_export_utils.py [hunk 1]
```diff
@@ -165,7 +165,7 @@ def base_common(self, row, type_name):
             "t_id": self.get_tid(row),
         }
 
-        if self.current_basket is not None:
+        if self.current_basket is not None and type_name != "organisation":
```
Comment (sjib) [inline qgepqwat2ili/utils/qgep_export_utils.py:168]: @domi4484 Here you are not writing the current_basket if you have class organisation - is this the problem. The tests fail on organisation data.

File: qgepqwat2ili/tests/test_qgep.py [hunk 2]
```diff
@@ -147,7 +147,7 @@ def test_case_d_import_complete_xtf_to_qgep(self):
         session.close()
 
     # test for VSA_KEK_2019_LV95 export with selection and labels
-    def test_case_e_export_selection(self):
+    def _test_case_e_export_selection(self):
```
Comment (domi4484) [inline qgepqwat2ili/tests/test_qgep.py:150]:
```suggestion
    @pytest.mark.skip(reason="Selection export broken with baskets")
    def test_case_e_export_selection(self):
```

File: qgepqwat2ili/tests/test_qgep.py [hunk 3]
```diff
@@ -214,7 +214,7 @@ def test_case_e_export_selection(self):
         )
 
     # test for SIA405_ABWASSER_2015_LV95 export with selection and labels
-    def test_case_f_export_selection_sia405(self):
+    def _test_case_f_export_selection_sia405(self):
```
Comment (domi4484) [inline qgepqwat2ili/tests/test_qgep.py:217]:
```suggestion
    @pytest.mark.skip(reason="Selection export broken with baskets")
    def test_case_f_export_selection_sia405(self):
```

**Comment by domi4484** [comment]
It seems when importing that two organisations VSA and SBU get created with an invalid id. This gives then a validation error on export.
@sjib I don't get why this is suddently happen. How should look the content of the table organistation after importing the xtf you gave me sometimes ago (transferdatensatz_sia405_abwasser_2015_d_LV95_20210517.xtf)? Do you understand what is the issue?

ksnip_20241211-144624

ksnip_20241211-144619

**Comment by ponceta** [comment]
VSA should be `ch080qwzPR000017` or `ch080qwzOG000017`
and 
SBU should be `ch080qwzPR000018` or `ch080qwzOG000018`

Or something with a valid prefix but not `00000000OGXXXXXX` where `00000000` is the default prefix value which is invalid.

**Comment by sjib** [comment]
> It seems when importing that two organisations VSA and SBU get created with an invalid id. This gives then a validation error on export. @sjib I don't get why this is suddently happen. How should look the content of the table organistation after importing the xtf you gave me sometimes ago (transferdatensatz_sia405_abwasser_2015_d_LV95_20210517.xtf)? Do you understand what is the issue?
> 
> ksnip_20241211-144624
> 
> ksnip_20241211-144619

@domi4484 That is the problem that I describe in issue https://github.com/QGEP/qgepqwat2ili/issues/154
First attempt to solve this is in https://github.com/QGEP/qgepqwat2ili/pull/156 (draft)

**Comment by domi4484** [comment]
@ponceta @sjib thanks for the infos. Still I don't get what has been modified to change the import behavior for tests.

@sjib I see a difference with issue #154. In my screenshot `identifier` is `VSA` and in your its the obj_id `ch080qwzPR000012`. So I searched a bit more and noticed that the metaattribute are different depending on the model. Is that to be expected?

In that case I guess the fix from #156 must be adapted and implemented for sia405 and kek as well?

image

**Comment by sjib** [comment]
@domi4484
In the releases 2015 of the datamodels `datenherr `and `datenlieferant `are still defined as text. So it is possible to just add a text value and not an obj_id. 

We have two different use cases:
Case a) you see in the import from SIA405 Abwasser 2015 uses text values 'VSA' and 'SBU'
Case b) you see in the import from VSA-DSS 2015 uses `ch080qwzPR000012` - an obj_id

That `datenherr `and `datenlieferant `are referenced to the class organisation was introduced with the model release 2020.

So to be able to migrate from 2015 to 2020 organisations specified as text in `datenherr `and `datenlieferant `must be also in the table organisation. On INTERLIS import 2015 this works already, see example above where based on the text values `VSA `and `SBU ` two new organisations are created.
So that is correct and as we expect it to be.

That they do not get a valid obj_id is due to the settings in `qgep_sys.oid_prefixes` with the prefix `000000 `beeing set to true. 
If you change the oid_prefixes to a value starting with `chxxxxxx `it will create a valid obj_id.

So my question is, whether we can change and need to change the selected oid_prefix for the tests? But I would treat this in a separate new issue if needed.

So I think in your developments in this pull request you should not invest time to check whether it is a valid obj_id but solve the t_basket problem. If this is solved - then ok and no further adaptions needed.

What needs to be solved is the case described in https://github.com/QGEP/qgepqwat2ili/issues/154

If we use an obj_id of an organisation in datenherr / datenlieferant that is also in the dataset as own organisation dataset, then it should not create an additional new organisation on the INTERLIS import - see https://github.com/QGEP/qgepqwat2ili/issues/154#issuecomment-2265950656

But I suggest also to treat this in a separate pull request to not mix issues.

Does this help?

Else contact me by email and we can have a short exchange online to further sort out the things.

**Comment by domi4484** [comment]
> On INTERLIS import 2015 this works already, see example above where based on the text values `VSA `and `SBU ` two new organisations are created.

In the import I don't see any code handling those values. Where is this implemented?



> If you change the oid_prefixes to a value starting with `chxxxxxx `it will create a valid obj_id.

So I guess for the tests I could just change the default value. Still I don't understand what makes this issue arise. In the main branch this is not happening.




> So I think in your developments in this pull request you should not invest time to check whether it is a valid obj_id but solve the t_basket problem. If this is solved - then ok and no further adaptions needed.

The baskets seems to be set correctly as the failing message is for SIA405. The KEK export ends without errors. So I can't really ignore that part.


To see how a correct export look like I checked out v0.14 but it seems broken. So I need to ask if the KEK export even worked in the past?

1. When kicking the export you always get a python error
image
2. If I fix that error I get the next one (for example when exporting the wincan test file (case_a_import_from_wincan.xtf)) about not finding some reference when exporting the class file / datei.
3. If I comment out the export for file / datei, the KEK export finishes without errors, but the KEK XTF file in the end is empty. (The SIA405 XTF is not empty but invalid)

**Comment by sjib** [comment]
> > On INTERLIS import 2015 this works already, see example above where based on the text values `VSA `and `SBU ` two new organisations are created.
> 
> In the import I don't see any code handling those values. Where is this implemented?

This is handeled here:
    def create_or_update_organisation(name):

  https://github.com/QGEP/qgepqwat2ili/blob/27ba9678e6fc56089e7b75622c602a2a7c44bef2/qgepqwat2ili/qgepsia405/import_.py#L90

**Comment by sjib** [comment]
> The baskets seems to be set correctly as the failing message is for SIA405. The KEK export ends without errors. So I can't really ignore that part.

Do you get check errors (from github) or ilivalidator errors?

**Comment by domi4484** [comment]
> > > On INTERLIS import 2015 this works already, see example above where based on the text values `VSA `and `SBU ` two new organisations are created.
> > 
> > 
> > In the import I don't see any code handling those values. Where is this implemented?
> 
> This is handeled here: def create_or_update_organisation(name):
> 
> https://github.com/QGEP/qgepqwat2ili/blob/27ba9678e6fc56089e7b75622c602a2a7c44bef2/qgepqwat2ili/qgepsia405/import_.py#L90

Ok but there is no special handling of the values "VSA" and "SBU" right? 2 "normals" organisation get created with the default prefix

**Comment by sjib** [comment]
> > > > On INTERLIS import 2015 this works already, see example above where based on the text values `VSA `and `SBU ` two new organisations are created.
> > > 
> > > 
> > > In the import I don't see any code handling those values. Where is this implemented?
> > 
> > 
> > This is handeled here: def create_or_update_organisation(name):
> > https://github.com/QGEP/qgepqwat2ili/blob/27ba9678e6fc56089e7b75622c602a2a7c44bef2/qgepqwat2ili/qgepsia405/import_.py#L90
> 
> Ok but there is no special handling of the values "VSA" and "SBU" right? 2 "normals" organisation get created with the default prefix

Yes, we have nothing hardcoded. Why should we?
But we can handle whether an organisation comes with its identifier such as  "`VSA`" and "`SBU`" or with its obj_id such as `ch080qwzPR000012`

**Comment by domi4484** [comment]
> > > > > On INTERLIS import 2015 this works already, see example above where based on the text values `VSA `and `SBU ` two new organisations are created.
> > > > 
> > > > 
> > > > In the import I don't see any code handling those values. Where is this implemented?
> > > 
> > > 
> > > This is handeled here: def create_or_update_organisation(name):
> > > https://github.com/QGEP/qgepqwat2ili/blob/27ba9678e6fc56089e7b75622c602a2a7c44bef2/qgepqwat2ili/qgepsia405/import_.py#L90
> > 
> > 
> > Ok but there is no special handling of the values "VSA" and "SBU" right? 2 "normals" organisation get created with the default prefix
> 
> Yes, we have nothing hardcoded. Why should we? But we can handle whether an organisation comes with its identifier such as "`VSA`" and "`SBU`" or with its obj_id such as `ch080qwzPR000012`

I am just trying to understand why those ID are making problems in this PR that in fact doesn't change anything in the import. And I think I found out now:

The test `test_case_f_export_selection_sia405` which is failing with the organisation ID error doesn't explicitly set up the database as other tests do.
- This means it uses the setup of another test.
- But test are not guaranteed to run in a special order -> this test is probably in this case not using the demodata as a base and for this reason VSA and SBU get created with an invalid ID

File: qgepqwat2ili/qgep/export.py [hunk 4]
```diff
@@ -21,6 +21,9 @@ def qgep_export_kek(selection=None, labels_file=None, orientation=None, basket_e
     Args:
         selection:      if provided, limits the export to networkelements that are provided in the selection
     """
+    logger.warning(
+        f"Starting qgep_export_kek: selection={selection}, labels_file={labels_file}, orientation={orientation}, basket_enabled={basket_enabled}",
+    )
```
Comment (domi4484) [inline qgepqwat2ili/qgep/export.py:26]:
```suggestion
```

File: qgepqwat2ili/qgepdss/export.py [hunk 5]
```diff
@@ -21,6 +21,9 @@ def qgep_export_dss(selection=None, labels_file=None, orientation=None, basket_e
     Args:
         selection:      if provided, limits the export to networkelements that are provided in the selection
     """
+    logger.warning(
+        f"Starting qgep_export_dss: selection={selection}, labels_file={labels_file}, orientation={orientation}, basket_enabled={basket_enabled}",
+    )
```
Comment (domi4484) [inline qgepqwat2ili/qgepdss/export.py:26]:
```suggestion
```

File: qgepqwat2ili/qgepsia405/export.py [hunk 6]
```diff
@@ -20,6 +20,9 @@ def qgep_export_sia405(selection=None, labels_file=None, orientation=None, baske
     Args:
         selection:      if provided, limits the export to networkelements that are provided in the selection
     """
+    logger.warning(
+        f"Starting qgep_export_sia405: selection={selection}, labels_file={labels_file}, orientation={orientation}, basket_enabled={basket_enabled}",
+    )
```
Comment (domi4484) [inline qgepqwat2ili/qgepsia405/export.py:25]:
```suggestion
```

File: qgepqwat2ili/tests/test_qgep.py [hunk 7]
```diff
@@ -147,6 +147,7 @@ def test_case_d_import_complete_xtf_to_qgep(self):
         session.close()
 
     # test for VSA_KEK_2019_LV95 export with selection and labels
+    @unittest.skip("Selection export broken with baskets")
```
Comment (domi4484) [inline qgepqwat2ili/tests/test_qgep.py:150]:
```suggestion
```

**Comment by sjib** [comment]
@domi4484 Here some testdata for importing VSA-KEK 2019 data:
[testdata_vsa_kek_2019.zip](https://github.com/user-attachments/files/18155962/testdata_vsa_kek_2019.zip)

**Comment by domi4484** [comment]
> I tried to test the current status - No output in KEK file:

Yes in the tests output it looks empty as well. 

Thanks for the additional testdata. I think I found the issue. The objects are put into the wrong basket and so ili2db can't export them. Ill try to fix that

**Comment by domi4484** [comment]
Now the .xtf file is filled with data for me locally. On the pipeline the file is still empty

**Comment by sjib** [comment]
@domi4484 Can you finish this today so we can make the release of 1.6.6 tomorrow?

**Comment by ponceta** [comment]
As this is working on local installations, we will merge it, test it and fix the CI in 2025.

**Pull request merged**